jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/dialects/impl/HiveSqlGenerationVisitor.java

```diff
@@ -0,0 +1,257 @@
+package com.exasol.adapter.dialects.impl;
+
+import com.exasol.adapter.dialects.SqlDialect;
+import com.exasol.adapter.dialects.SqlGenerationContext;
+import com.exasol.adapter.dialects.SqlGenerationVisitor;
+import com.exasol.adapter.jdbc.ColumnAdapterNotes;
+import com.exasol.adapter.metadata.ColumnMetadata;
+import com.exasol.adapter.sql.*;
+import com.google.common.base.Joiner;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by np on 12/30/2016.
+ */
+public class HiveSqlGenerationVisitor extends SqlGenerationVisitor {
+
+
+    public HiveSqlGenerationVisitor(SqlDialect dialect, SqlGenerationContext context) {
+        super(dialect, context);
+    }
+
+
+    @Override
+    public String visit(SqlSelectList selectList) {
+        List<String> selectListElements = new ArrayList<>();
+        SqlStatementSelect select = (SqlStatementSelect) selectList.getParent();
+        if (selectList.isSelectStar()) {
+            if (selectListRequiresCasts(selectList)) {
+                // Do as if the user has all columns in select list
+                int columnId = 0;
+                for (ColumnMetadata columnMeta : select.getFromClause().getMetadata().getColumns()) {
+                    SqlColumn sqlColumn = new SqlColumn(columnId, columnMeta);
+                    if (columnMeta.getType().toString().equals("VARCHAR(2000000) UTF8")) {
+                        selectListElements.add("base64(" + super.visit(sqlColumn) + ")");
+                    } else {
+                        selectListElements.add(super.visit(sqlColumn));
+                    }
+                    ++columnId;
+
+                }    }
+            else {
+                    selectListElements.add("*");
+                }
+
+        } else {
+            if(selectList.isRequestAnyColumn()){
+                return "1";
+            }
+            List<String> binaryElements = new ArrayList<>();
+            int columnId = 0;
+            for (ColumnMetadata columnMeta : select.getFromClause().getMetadata().getColumns()) {
+                SqlColumn sqlColumn = new SqlColumn(columnId, columnMeta);
+                String typeName = ColumnAdapterNotes.deserialize(sqlColumn.getMetadata().getAdapterNotes(), sqlColumn.getMetadata().getName()).getTypeName();
+                if(typeName.equals("BINARY")  ){
+                    binaryElements.add(super.visit(sqlColumn));
+                }
+                ++columnId;
+            }
+                for (SqlNode node : selectList.getExpressions()) {
+                    if (binaryElements.contains(node.accept(this))) {
+                        selectListElements.add("base64(" + node.accept(this) + ")");
+                    } else {
+                        selectListElements.add(node.accept(this));
+                    }
+                }
+            }
```

Comment:
Now I understand why you go through the from clause first. I think we should do a final small refactoring to make it more simple. You can directly iterate over the selectList and then check for each node if it is is a SqlColumn. If yes, we can obtain the metadata and the adapterNotes from the SqlColumn and check if it is of type binary. We should avoid string matching whenever possible.